---
title: "Metabarcoding primer evaluation"
subtitle: "Pest list assembly"
author: "Alexander Piper"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
# Knitr global setup - change eval to true to run code
library(knitr)
knitr::opts_chunk$set(echo = TRUE, eval=FALSE, message=FALSE,error=FALSE, fig.show = "hold", fig.keep = "all")
opts_chunk$set(dev = 'png')
```

## Load packages
```{r load packages}
#Set required packages
.cran_packages <- c("usethis",
                    "tidyverse",
                    "spider", 
                    "insect",
                    "ape",
                    "RColorBrewer",
                    "seqinr",
                    "patchwork",
                    "ShortRead",
                    "foreach",
                    "doParallel",
                    "TmCalculator",
                    "castor",
                    "furrr",
                    "UpSetR",
                    "DescTools")

.bioc_packages <- c("DECIPHER",
                    "ggtree",
                    "Biostrings")

.inst <- .cran_packages %in% installed.packages()
if(any(!.inst)) {
   install.packages(.cran_packages[!.inst])
}
.inst <- .bioc_packages %in% installed.packages()
if(any(!.inst)) {
  if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
  BiocManager::install(.bioc_packages[!.inst], ask = F)
}

#Load all packages
sapply(c(.cran_packages,.bioc_packages), require, character.only = TRUE)

# Github packages
devtools::install_github("alexpiper/taxreturn")
library(taxreturn)
library(PrimerMiner)

# SOurce internal functions
source("R/helper_functions.R")
source("R/themes.R")
```

# Assemble pest list

* EPPO global database https://gd.eppo.int/ - DONE
* US APHIS - https://www.aphis.usda.gov/aphis/home/ - DONE
* QBank - https://qbank.eppo.int/arthropods/organisms - DONE
* Global invasive species database - http://www.iucngisd.org/gisd/search.php - DONE
* Global register of introduced or invasive species http://www.griis.org/ - DONE
* VectorBase: https://www.vectorbase.org/organisms - DONE
* DAWR top 40 - http://www.agriculture.gov.au/pests-diseases-weeds/plant - DONE
* PHA National biosecurity status report -  http://www.planthealthaustralia.com.au/national-programs/national-plant-biosecurity-status-report/ - DONE
* Ashfaq & Herbert 2016 - DNA barcodes for bio-surveillance: regulated and economically important arthropod plant pests - DONE
* CABI - https://t.co/LGjlFoOazd - DONE
* http://www.europe-aliens.org - DONE

```{r Curate pest lists}
dat <- list.files("primer_evaluation/pestlist/", pattern = ".csv", full.names = TRUE) %>%
  purrr::set_names() %>%
  map_dfr(read_csv, .id = "Source", col_types = cols("Species" = col_character())) %>%
  mutate(Source = str_remove(basename(Source), pattern="\\.csv")) %>%
  mutate(Species = Species %>%  
           str_remove_all("Ã¿") %>% #Resolve weird characters
           iconv(from = 'UTF-8', to = 'ASCII//TRANSLIT')%>% 
           str_remove_all("\\?") %>%
           str_remove_all("\\((.*?)\\)") %>% # remove things between brackets ie: Hygromia (Hygromia) cinctella
           str_squish() # remove excess whitespace
         ) %>%
  filter(str_count(Species, " ") > 0 ) %>% #Remove non-binomial 
  separate(Species, into=c("Genus", "Species"), sep=" ", extra="merge") %>% # Fix duplicated genus names
  mutate(Species = str_remove(Species, pattern=Genus) %>% str_squish()) %>%
  unite(col=Species, Genus, Species, sep = " ") %>%
  unique()

# Map to OTT taxonomy ids
db <- taxreturn::get_ott_taxonomy(dir="ott3.2", filter_unplaced = FALSE)

dat_resolved <- dat %>% mutate(mapped =taxreturn::map_to_ott(Species, db, resolve_synonyms=TRUE,  dir="ott3.2", filter_unplaced=TRUE, remove_na = FALSE, quiet=FALSE) ) %>%
  mutate(Species = mapped %>% str_remove("^.*;"),
         taxid = mapped %>% str_remove(";.*$") %>% str_remove("^.*\\|") )%>%
  mutate(taxid = na_if(taxid, "NA")) %>%
  dplyr::filter(!is.na(taxid)) %>%
  rownames_to_column("rows") %>%
  mutate(mapped = mapped %>% str_remove("^NA") %>% paste0(rows, .)) %>%  #Add dummy accession number
  dplyr::select(-rows)

lineage <- get_ott_lineage(dat_resolved$mapped, db) %>%
  bind_cols(dat_resolved) %>% 
  dplyr::select(-mapped, -tax_name, -Species) %>% 
  rename_all(funs(str_to_sentence(.))) %>%
  filter(Class %in% c("Insecta", "Arachnida")) %>%
  drop_na()

#Write out final list of pests 
write_csv(lineage, "primer_evaluation/pestlist.csv")
```

## Seqs per pest taxa

```{R}
seqs <- readFASTA("reference/database_builder/09_lengthfilt.fa.gz")
pestlist <- read_csv("primer_evaluation/pestlist.csv") %>%
  filter(Class=="Insecta") %>%
  distinct()

taxa <- names(seqs)%>%
  enframe() %>%
  select(-name) %>%
  separate(value, into=c("acc", "species"), sep=";") %>%
  mutate(pest = case_when(
    species %in% pestlist$Species ~ TRUE,
    !species %in% pestlist$Species ~ FALSE
  ))

#Number of taxa per species
n_seqs <- taxa %>% 
  group_by(species, pest) %>%
  summarise(sum = n())

# Mean number of seqs per spp
t.test(n_seqs %>% filter(pest) %>% pull(sum), n_seqs %>% filter(!pest) %>% pull(sum)) %>%
  broom::tidy()

# Summary of seqs per pest or non-pest species
n_seqs %>%
  group_by(pest) %>%
  summarise(mean = mean(sum), total = sum(sum), lower = range(sum)[1], upper = range(sum)[2], se=sqrt(var(sum)/length(sum)))

# How many taxa didnt have seqs
table(pestlist$Species %in% taxa$species)

# Summary of seqs per all species
n_seqs %>%
  ungroup() %>%
  summarise(mean = mean(sum), total = sum(sum), lower = range(sum)[1], upper = range(sum)[2], se=sqrt(var(sum)/length(sum)))
  

```

# Summarise how many actually in database

```{r}
pestlist <- read_csv("primer_evaluation/pestlist.csv") %>%
  filter(Class=="Insecta") %>%
  distinct()

test <- pestlist %>%
  dplyr::select(-Acc) %>%
  left_join(
    taxa  %>%
    filter(pest) %>%
    tidyr::separate(acc, into=c("acc", "Taxid"), sep="\\|") %>%
      mutate(Taxid = as.numeric(Taxid)) %>%
      select(Species = species, -Taxid, pest, acc)) %>%
  mutate(present = case_when(
    !is.na(pest) ~ TRUE,
    is.na(pest) ~ FALSE
  ))

test %>% 
  group_by(present) %>%
  summarise(Order = n_distinct(Order), Genus = n_distinct(Genus), Species = n_distinct(Species), seqs = n_distinct(acc))
  

  
```


## Supplementary figure

```{r plot}
lineage <- read_csv("primer_evaluation/pestlist.csv") %>%
  distinct() %>%
  mutate(Source = Source %>% str_remove("export_")) %>%
  mutate(Source = case_when(
    Source == "griis" ~ "GRIIS",
    Source=="gisd" ~ "GISD",
    Source=="europealiens" ~ "DAISIE",
    Source=="eppo" ~ "EPPO",
    Source=="ashfaq" ~ "Ashfaq et al 2016",
    Source=="qbank" ~ "Q-Bank",
    Source=="cabi" ~ "CABI",
    Source=="pha" ~ "PHA",
    Source=="vectorbase" ~ "VectorBase",
    Source=="dawr40" ~ "DAWR Top 40"
  ))


library(ggupset)
gg.upset <- lineage %>%
  dplyr::select(Species, Source) %>%
  group_by(Species) %>%
  summarise(Source = list(Source)) %>%
  ggplot(aes(x=Source)) +
    geom_bar() +
    scale_x_upset(n_intersections = 25) +
    base_theme +
    theme(panel.grid = element_blank()) +
  labs(x = "Intersection between datasets", y="Intersection size")

gg.upset

# Total set size
gg.total <- lineage %>%
  mutate(Source = factor(Source, levels=rev(c("GRIIS", "DAISIE", "EPPO", "Ashfaq et al 2016", "Q-Bank", "CABI", "PHA", "VectorBase", "DAWR Top 40", "GISD")))) %>%
  mutate(label = case_when(
   Class == "Insecta" & Order %in% c("Diptera" ,"Coleoptera", "Hemiptera", "Hymenoptera", "Lepidoptera", "Orthoptera") ~ Order,
  Class == "Arachnida" ~ "Arachnida",
  Class == "Insecta" & !Order %in% c("Diptera" ,"Coleoptera", "Hemiptera", "Hymenoptera", "Lepidoptera", "Orthoptera") ~ "Other Insects"
  )) %>%
  ggplot(aes(x=Source, fill=label))+
  geom_bar() +
  scale_fill_brewer(palette="Paired") + 
  base_theme +
  labs(x=NULL, y="Total records", fill="Taxon") +
  coord_flip()+
  theme(legend.position = "bottom") +
  scale_y_reverse()
  
# PCA of reference database
#Use unifrac and phylogenetic tree?


pca_dat <- lineage %>%
  dplyr::select(Species, Source) %>%
  distinct() %>%
  mutate(value = 1) %>%
  pivot_wider(names_from = Source, values_from=value, values_fill=0) %>%
  column_to_rownames("Species")


database_pca <- prcomp(vegan::vegdist(t(pca_dat), method="jaccard" ), scale. = TRUE)

gg.database_pca <-database_pca %>%
  broom::augment() %>%
    ggplot(aes(.fittedPC1, .fittedPC2, colour=.rownames, fill=.rownames)) +
    geom_point(size = 4, alpha = 0.8, shape=21, colour="black") +
  ggrepel::geom_text_repel(aes(label = .rownames, colour=.rownames))+
    labs(x = paste0("Principal component 1 (",  percent(percent_variation[1]), ")"), 
         y = paste0("Principal component 2 (",  percent(percent_variation[2]),")")) +
  geom_vline(xintercept = 0)+
  geom_hline(yintercept = 0) +
  base_theme +
  scale_color_brewer(palette = "Paired") +
  scale_fill_brewer(palette = "Paired") 

# Assemble figure 1

Fig1a <- gg.database_pca / gg.total + plot_layout(heights = c(2,1))

Fig1b <- gg.upset / grid::textGrob('')+ plot_layout(heights = c(10,1))

Fig1 <- Fig1a - Fig1b +  plot_annotation(tag_levels = 'A') 

Fig1

#Save figure 1
pdf(file="fig/Fig1_pestlist_summary.pdf", width = 11, height = 8 , paper="a4r")
  plot(Fig1)
try(dev.off(), silent=TRUE)
  
# Summaries for article text

# Unique taxa
lineage %>% 
  #select(-Source, Acc) %>%
  summarise(Species = n_distinct(Species),
            Genus = n_distinct(Genus),
            Family = n_distinct(Family),
            Order = n_distinct(Order),
            )
# Sum of reference DB's
lineage %>% 
  group_by(Source) %>%
  summarise(Species = n_distinct(Species)) %>%
  arrange(Species)

# Proportion of sequences unique
lineage %>%
  add_count(Source, name = "DB_total") %>%
  group_by(Species) %>%
  add_tally(name = "n_occurances") %>%
  ungroup() %>%
  filter(n_occurances==1) %>%
  group_by(Source, DB_total)%>%
  summarise(n = n_distinct(Species)) %>%
  mutate(freq = n / DB_total)%>%
  arrange(freq)
```